soc: nrf: Use data from DTS to populate HAS_HW_NRF_* Kconfig options

anangl opened the pull request:
Instead of selecting appropriate HAS_HW_NRF_* options for particular
nRF SoCs (and simulated nRF52 target), set their values basing on
information from devicetree.
Correct also semantics of those options so that they are set only when
a corresponding DT node is enabled. This allows using them directly in
Kconfig dependencies of Zephyr drivers for nRF peripherals. Update
appropriately these dependencies.

Refactor also nrfx related Kconfig options, as they used HAS_HW_NRF_*
options in their dependencies.
Change dependencies of the Kconfig options that enable particular nrfx
drivers so that the drivers and their instances are available when a
devicetree node with a corresponding compatible property exists, no
matter if it is enabled or not.
Change the options that enable compilation of multi-instance drivers
to be promptless helpers only (not directly configurable for users),
as they need to be enabled only together with some options that enable
particular instances of a given driver, and those options select them
as needed.

anangl committed:
scripts: kconfigfunctions: Add dt_has_compat()

Add a function for checking if any node with a given compatible exists,
no matter if its status is "okay" or not.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
boards: nrf51: Suppress DTC warnings about duplicate unit-address

Prevent "unique_unit_address_if_enabled" warnings from being reported
for nRF51 Series SoCs, where certain nodes need to be enabled with
the same base addresses. These are:
- power@40000000 & clock@40000000 & mpu@40000000

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
dts: Add missing nodes and bindings for peripherals present in nRF SoCs

Add missing devicetree nodes representing the following peripherals
in nRF SoCs:
- ACL
- BPROT
- CCM
- COMP/LPCOMP
- CTRLAP
- DCNF
- MPU (nRF MPU peripheral in nRF51 Series, not ARM MPU)
- MUTEX
- MWU
- NFCT
- OSCILLATORS
- POWER (in nRF51 and nRF52 Series)
- PPI
- RESET
- SWI
- USBREG
Add also corresponding bindings and validation of base addresses of
these nodes.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
dts: nordic: Add properties indicating certain peripheral features

Add boolean properties to the binding for the nRF RADIO peripheral
to indicate whether the following features are supported:
- IEEE 802.15.4 mode
- 2 Mbps BLE mode
- coded BLE PHY
- high TX power settings
Set these properties appropriately in devicetree radio nodes for all
nRF SoCs.

Add also such properties and set them in appropriate nodes for nRF
flash controllers to indicate whether they support partial erase.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
scripts: kconfigfunctions: Redefine dt_nodelabel_has_compat()

The function in its current form is confusing because unlike other
similarly named functions (dt_nodelabel_has_prop(), dt_node_has_prop())
or devicetree macros (DT_NODE_HAS_COMPAT(), DT_NODE_HAS_PROP()), this
function takes into account the status of the checked node and returns
"y" only when the node is enabled.
This commit redefines dt_nodelabel_has_compat() so that it no longer
checks the node status, and for cases where the previous functionality
is needed, a new function named dt_nodelabel_enabled_with_compat()
is introduced as a replacement.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
modules: hal_nordic: Refactor nrfx related Kconfig options

Change dependencies of the Kconfig options that enable particular nrfx
drivers so that the drivers and their instances are available when
a devicetree node with a corresponding compatible property exists,
no matter if it is enabled or not.
Change the options that enable compilation of multi-instance drivers
to be promptless helpers only (not directly configurable for users),
as they need to be enabled only together with some options that enable
particular instances of a given driver, and those options select them
as needed.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
soc: nrf: Use data from DTS to populate HAS_HW_NRF_* Kconfig options

Instead of selecting appropriate HAS_HW_NRF_* options for particular
nRF SoCs (and simulated nRF52 target), set their values basing on
information from devicetree.
Correct also semantics of those options so that they are set only when
a corresponding DT node is enabled. This allows using them directly in
Kconfig dependencies of Zephyr drivers for nRF peripherals. Update
appropriately these dependencies.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
boards: nrf52: Suppress DTC warnings about duplicate unit-address

Prevent "unique_unit_address_if_enabled" warnings from being reported
for nRF52 Series SoCs, where certain nodes need to be enabled with
the same base addresses. These can be (depending on a given SoC):
- power@40000000 & clock@40000000
- power@40000000 & clock@40000000 & bprot@40000000
- acl@4001e000 & flash-controller@4001e000

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
dts: nrf5340: Use dual compatible property for mbox nodes

This is a follow-up to commit cf6a58d3f61aa163d149a3c3ff6389285bd09c13.

Restore the "nordic,nrf-ipc" compatible property in mbox nodes for both
nRF5340 cores and use it together with the new "nordic,mbox-nrf-ipc"
one. This way either the MBOX or the IPM driver can be used for these
nodes without further modifications. This eliminates the need to use
overlays in quite a few cases, so remove all those no longer needed
ones (which are also a bit confusing now as they refer to no longer
existing ipc@2a000 and ipc@41012000 nodes).

Restore also the ipc node label removed in the commit mentioned above,
as the label is used in validation of base addresses of nRF DT nodes.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
Rebased onto #44306.

mbolivar-nordic commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -596,7 +597,6 @@ def shields_list_contains(kconf, _, shield):
         "dt_node_int_prop_hex": (dt_node_int_prop, 2, 3),
         "dt_node_str_prop_equals": (dt_node_str_prop_equals, 3, 3),
         "dt_node_has_compat": (dt_node_has_compat, 2, 2),
-        "dt_nodelabel_has_compat": (dt_nodelabel_has_compat, 2, 2),
```
I don't think we make any API stability guarantees about these functions, but did you consider printing a deprecation warning if the function is called under the old name, instead of removing it entirely? My concern here is about out of tree users.

mbolivar-nordic commented on modules/hal_nordic/nrfx/Kconfig:
```diff
@@ -76,172 +74,168 @@ config NRFX_GPIOTE_NUM_OF_EVT_HANDLERS
+	select NRFX_USBREG if $(dt_has_compat,$(DT_COMPAT_NORDIC_NRF_USBREG))
 
 config NRFX_PPI
 	bool "PPI allocator"
-	depends on HAS_HW_NRF_PPI
+	# depends on $(dt_has_compat,$(DT_COMPAT_NORDIC_NRF_PPI))
```
I guess this is commented out because it is still a draft?

anangl replied:
To some extent. :wink: I actually forgot to restore this dependency. Thanks for spotting it!

mbolivar-nordic reviewed (commented):
Looks great! I will try to double-check in more detail when it is out of draft.

anangl committed:
dts: nrf: Rename "nordic,nrf-watchdog" compatible to "nordic,nrf-wdt"

For consistency with the scheme used for other nRF peripherals,
use the peripheral name that nRF Product Specifications use.
In this case, it is WDT, not WATCHDOG.

Also remove the requirement for the label property in the binding.
It is no longer needed.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

joerchan commented on soc/arm/nordic_nrf/Kconfig.peripherals:
```diff
@@ -4,305 +4,298 @@
 # SPDX-License-Identifier: Apache-2.0
 
 config HAS_HW_NRF_ACL
```
Are we keeping these for compatibility, or should they be removed?

anangl replied:
We are keeping them, as they are useful for creating Kconfig dependencies for drivers or other modules that need particular hardware or feature.

joerchan replied:
But can't they use `$(dt_compat_enabled,$(DT_COMPAT_NORDIC_NRF_ACL))` instead?
Now we have 2 ways to do the same thing, I get that this removes the pattern for having select, but I don't really understand why you would do this change for example:
```
config NFCT_PINS_AS_GPIOS
	bool "NFCT pins as GPIOs"
-	depends on HAS_HW_NRF_NFCT
+	depends on $(dt_has_compat,$(DT_COMPAT_NORDIC_NRF_NFCT))
```

anangl replied:
> But can't they use `$(dt_compat_enabled,$(DT_COMPAT_NORDIC_NRF_ACL))` instead?

They could. `HAS_HW_NRF_*` are just more convenient to use. Plus you can refer to `CONFIG_HAS_HW_NRF_*` definitions from source code (we've got a few such uses in tree).

> Now we have 2 ways to do the same thing, I get that this removes the pattern for having select, but I don't really understand why you would do this change for example:
> [..]

`HAS_HW_NRF_NFCT` is set to "y" only when a `"nordic,nrf-nfct"` compatible node is enabled, and the `NFCT_PINS_AS_GPIOS` option needs to be available when such node just exists (i.e. when a given SoC supports NFCT). That's why I needed to do this change.

joerchan replied:
Ok, thank you for the clarification. I see now that you have actually changed the semantics of the `CONFIG_HAS_HW_NRF_*` from the HW being supported by the SoC to the HW being enabled in devicetree.

So `$(dt_compat_enabled,$(DT_COMPAT_NORDIC_NRF_ACL))` represents the old semantics, then, correct?

anangl replied:
> So `$(dt_compat_enabled,$(DT_COMPAT_NORDIC_NRF_ACL))` represents the old semantics, then, correct?

No. That would need to be `$(dt_has_compat,$(DT_COMPAT_NORDIC_NRF_ACL))`. Such dependencies are now used in `modules/hal_nordic/nrfx/Kconfig` instead of `HAS_HW_NRF_*` symbols used previously, because nrfx drivers need to be available when corresponding HW is supported by a given SoC, not only when it is enabled to be used by Zephyr drivers.

gmarull commented on doc/guides/build/kconfig/preprocessor-functions.rst:
```diff
@@ -28,6 +28,7 @@ while the ``*_hex`` version returns a hexadecimal value starting with ``0x``.
 
 .. code-block:: none
 
+   $(dt_has_compat,<compatible string>)
```
maybe `dt_compat_exists`?

anangl replied:
I based on the work started in #35398 and the `dt_has_compat` name comes from there. And to me it seems to suit better, as it is consistent with [`DT_HAS_CHOSEN`](https://github.com/zephyrproject-rtos/zephyr/blob/65a2e4b7600845f6fdef18233554e6a9b0e51d04/include/devicetree.h#L1861) and [`DT_HAS_COMPAT_STATUS_OKAY`](https://github.com/zephyrproject-rtos/zephyr/blob/65a2e4b7600845f6fdef18233554e6a9b0e51d04/include/devicetree.h#L2189). @mbolivar-nordic ?

gmarull replied:
makes sense

gmarull commented on modules/hal_nordic/nrfx/Kconfig:
```diff
@@ -9,62 +9,60 @@ menu "nrfx drivers"
 
 config NRFX_ADC
 	bool "ADC driver"
-	depends on HAS_HW_NRF_ADC
+	depends on $(dt_has_compat,$(DT_COMPAT_NORDIC_NRF_ADC))
```
I think it should be drivers that select HAL options, see for example ST/GD:

https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/hal_gigadevice/Kconfig#L230-L234

Then, the UART driver just does:

https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/serial/Kconfig.gd32#L7-L15

This results in:

- HAL modules being compiled as needed by drivers
- Drivers being automatically selected based on DT status

gmarull replied:
ok, got it, thanks for the explanation

anangl commented:
The CI failures are not related to changes introduced by this PR. I created #44453 for these failures.

anangl commented:
#44306 needs to go in first.

gmarull commented on boards/arm/96b_carbon_nrf51/pre_dt_board.cmake:
```diff
@@ -0,0 +1,6 @@
+
+# Suppress "unique_unit_address_if_enabled" to handle the following overlaps:
+# - power@40000000 & clock@40000000 & nrf-mpu@40000000
+list(APPEND EXTRA_DTC_FLAGS "-Wno-unique_unit_address_if_enabled")
```
note: would be nice to set this at SoC level if possible (one less thing to worry when creating a new board)
